finality-aleph/src/justification/scheduler.rs

```diff
@@ -0,0 +1,89 @@
+use crate::{MillisecsPerBlock, SessionPeriod};
+use std::{
+    cmp::min,
+    time::{Duration, Instant},
+};
+
+pub enum SchedulerActions {
+    ClearQueue,
+    Request,
+    Wait,
+}
+
+/// Bunch of methods for managing frequency of sending justification requests.
+pub trait JustificationRequestScheduler {
+    /// Decides whether we can request new justification.
+    fn schedule_action(&mut self) -> SchedulerActions;
+    /// Notice block finalization.
+    fn on_block_finalized(&mut self);
+    /// Notice request sending.
+    fn on_request_sent(&mut self);
+}
+
+pub struct JustificationRequestSchedulerImpl {
+    last_request_time: Instant,
+    last_finalization_time: Instant,
+    delay: Duration,
+    attempt: u32,
+    max_attemps: u32,
+}
+
+impl JustificationRequestSchedulerImpl {
+    pub fn new(
+        session_period: &SessionPeriod,
+        millisecs_per_block: &MillisecsPerBlock,
+        max_attemps: u32,
+    ) -> Self {
+        Self {
+            last_request_time: Instant::now(),
+            last_finalization_time: Instant::now(),
+            delay: Duration::from_millis(min(
+                millisecs_per_block.0 * 2,
+                millisecs_per_block.0 * session_period.0 as u64 / 10,
+            )),
+            attempt: 0,
+            max_attemps: max_attemps,
+        }
+    }
+}
+
+impl JustificationRequestSchedulerImpl {
+    fn enough_time_elapsed(&self) -> bool {
+        let now = Instant::now();
+
+        now - self.last_finalization_time > self.delay
+            && now - self.last_request_time > 2 * self.delay
+    }
+}
+
+impl JustificationRequestScheduler for JustificationRequestSchedulerImpl {
+    fn schedule_action(&mut self) -> SchedulerActions {
+        let now = Instant::now();
+        if self.enough_time_elapsed() {
+            if now - self.last_finalization_time > now - self.last_request_time {
+                self.attempt += 1;
+
+                if self.attempt == self.max_attemps
+                    && now - self.last_finalization_time > self.max_attemps * self.delay
```

Comment:
I think I get the logic here, but seems very convoluted. Maybe you could just clear `self.attempt` in the `on_block_finalized` method? The whole logic here would be "if attempts too high clear". with no need for all the timing conditions.

Reply:
Then you would end up with clearing a queue with every call of `schedule_action` after `self.attempt` reached some level. I'm not sure if I would like to do such logic and I don't know how to make current less convoluted. Maybe some comments would help?

Reply:
Ah, I meant this would be "attempts since last finalization or queue clearing", so you also reset the attempts when cleaning the queue ofc. That would work, wouldn't it?